rfc/0001-shapely-2.0-roadmap.md

```diff
@@ -0,0 +1,441 @@
+...
+TypeError: object of type 'Point' has no len()
+
+# MultiPolygon gets splitted
+>>> from shapely.geometry import MultiPolygon, box
+>>> mpoly = MultiPolygon([box(0, 0, 1, 1), box(2, 2, 3, 3)])
+>>> np.array([mpoly], dtype=object)
+array([[<shapely.geometry.polygon.Polygon object at 0x7f9846379ac0>,
+        <shapely.geometry.polygon.Polygon object at 0x7f9846379670>]],
+      dtype=object)
+```
+
+When working with arrays of geometry objects, the "iterability" of the
+geometries becomes an annoyance.
+
+The "Multi" geometry types also have a `.geoms` attribute, which gives access to
+a list of geometries to iterate over. So the question here can be: isn't this
+sufficient? Accessing this specific attribute doesn't seem problematic, as you
+need to know anyway that you have a "Multi" geometry, as the others will raise an
+error on iter/getitem.
+
+**Proposal for Shapely 2.0** Deprecate and eventually remove the iterability of
+"Multi" geometries, and point users to the `.geoms` attribute.
+
+This means that code like `mpoly[0]` or `for geom in mpoly: ...` needs to be
+replaced with `mpoly.geoms[0]` or `for geom in mpoly.geoms: ...`.
+
+
+### Empty geometries
+
+TO FILL IN
+
+Shapely 1.x is inconsistent in treating empty geometries, cfr https://github.com/Toblerity/Shapely/issues/742.
+
+
+### Spatial indexing with STRTree
+
+TO FILL IN
+
+PyGEOS also implements an STRtree. However, the API is not compatible with the
+one that is now in Shapely (but more similar to the API of `rtree`): the tree in
+PyGEOS returns integer indices, while the tree of Shapely returns actual
+geometry objects (cfr https://github.com/Toblerity/Shapely/issues/615).
+
+The behaviour of Shapely can easily be implemented on top of the behaviour of
+PyGEOS (to avoid a duplicate tree implementation). But the question here will
+still be how we want to provide this in the public API. Can we somehow unify
+this? Or just provide two tree objects users can choose from? Or is it
+acceptable to deprecate the current shapely behaviour? (or provide a keyword to
+still return geometries instead of integers?)
+
+
+### Prepared Geometries
+
+TO FILL IN
+
+In Shapely, there is a `PreparedGeometry` (and `prep` function), which needs to
+be used explicitly by the user. In PyGEOS, we are however thinking to "just" use
+this when appropriate under the hood, without the user needing to think about
+this. For example the STRtree already uses this (when a predicate is specified),
+and in https://github.com/pygeos/pygeos/pull/92 I am looking at enabling this in
+the vectorized spatial predicates like `contains` etc. The question here is if
+we will always know when this is "appropriate", or if there are cases that it
+can unexpectedly give a large overhead for situations prepared geoms were used
+unnecessarily (feedback on this in the linked PR is very welcome!). But if we do
+it under the hood, this could also mean that `PreparedGeometry` could be
+deprecated.
```

Comment:
My sense is that preparing all geometries by default in an effort to be user friendly goes too far (like the project did with the object-storing strtree). If you don't need it, you don't need it, and users and applications that do need it can just call an API function to prepare the geometry.